ppl/zqd/apiserver/manager.go

```diff
@@ -306,16 +311,29 @@ func (m *Manager) rowToSpaceInfo(ctx context.Context, sr schema.SpaceRow) (api.S
 	return spaceInfo, nil
 }
 
-func (m *Manager) GetSpace(ctx context.Context, id api.SpaceID) (api.SpaceInfo, error) {
+func (m *Manager) getSpacePermCheck(ctx context.Context, id api.SpaceID) (schema.SpaceRow, error) {
 	sr, err := m.db.GetSpace(ctx, id)
+	if err != nil {
+		return schema.SpaceRow{}, err
+	}
+	ident := auth.IdentifyFromContext(ctx)
+	if sr.TenantID != ident.TenantID {
+		return schema.SpaceRow{}, zqe.ErrForbidden()
```

Comment:
I don't understand what you mean about the different return paths?

If the `m.db.GetSpace` call fails, it's because of some database fetching error; there's no permissions checking that happens inside the db code. This function `getSpacePermCheck` is a convenience function to add permissions checking around the db GetSpace call, since it's needed in several places in Manager.

Reply:
OK, so the information leak is that returning from line 317 indicates that a space cannot be retrieved from the db (which may be because the space does not exist) while returning from line 321 indicates that the space exists and is owned by another tenant. It will leak information in the case that an unauthorized tenant acquires a valid space id that they do not own. Could you reverse the order of the m.db.GetSpace() and auth.IdentifyFromContext() calls and have it still work the way you want?

Reply:
> OK, so the information leak is that returning from line 317 indicates that a space cannot be retrieved from the db (which may be because the space does not exist) while returning from line 321 indicates that the space exists and is owned by another tenant. It will leak information in the case that an unauthorized tenant acquires a valid space id that they do not own. Could you reverse the order of the m.db.GetSpace() and auth.IdentifyFromContext() calls and have it still work the way you want?

I don't see how - we don't know the tenant id of the space until we retrieve it from the database.

Reply:
Yes, that is true. So by returning an error from the db, the code reveals the existence, non-existence, or any other possible db error state prior to knowing whether the tenant issuing the query is the correct tenant. That is why secure code generally uses "unauthorized" responses until authorization had been proven, regardless of the reason.